File: client/test/components/filters/PriceFilter.js
import expect from 'expect';
import React from 'react';
import Rcslider from 'rc-slider';
import { shallow, mount } from 'enzyme';
import Immutable from 'immutable';
import { FormControl } from 'react-bootstrap';

import PriceFilter from '../../../app/bundles/AnanasShop/components/filters/PriceFilter';

describe('component', () => {
  describe('PriceFilter', () => {
    let params;
    beforeEach( () => {
      params = {
        setFilter:      expect.createSpy(),
        updateFilter:   expect.createSpy(),
        fetchData:      expect.createSpy(),
        filter: Immutable.fromJS({
          type: 'PriceFilter',
          name: 'price',
          params: [],
          min: 0, max: 100, minB: 0,
          maxB: 100, minFormB: 0, maxFormB: 100
        })
      }
    })
    it('should render Rcslider', () => {
      const wrapper = shallow(<PriceFilter {...params}/>);
      expect(wrapper.find(Rcslider).length).toEqual(1);
    })

    it('should call fetchData onAfterChange event', () => {
      const wrapper = shallow(<PriceFilter {...params}/>);
      let rcSlider = wrapper.find(Rcslider);
      rcSlider.simulate('afterChange');
      expect(params.fetchData.calls.length).toEqual(1);
    })

    it('should call setFilter then Rcslider changed', () => {
      const wrapper = shallow(<PriceFilter {...params}/>);
      let rcSlider = wrapper.find(Rcslider);
      rcSlider.simulate('change', [0, 100]);

      expect(params.setFilter.calls[0].arguments)
        .toEqual([params.filter]);
    })

    it('should call updateFilter then form is onBlur', () => {
      const wrapper = shallow(<PriceFilter {...params}/>);
      let form = wrapper.find(FormControl).first();
      form.simulate('blur');

      expect(params.updateFilter.calls.length).toEqual(1);
      expect(params.updateFilter.calls[0].arguments)
        .toEqual([params.filter.set('params', Immutable.List([0,100]))]);
    })
  })
})

File: app/models/filters/price_filter.rb
class PriceFilter < BaseFilter
  def price
    price = params[:price]
    return [min, max] unless price.present?
    price.split(';')
  end

  def fields
    params[:fields].present? ? params[:fields].split(';') : []
  end

  def fields_products
    if params[:fields].present?
      @filter.products.joins( products_fields: :fields_values).where(fields_values: { id: fields }).uniq
    else
      @filter.products
    end
  end

  def min_b
    price[0].to_i
  end

  def max_b
    price[1].to_i
  end


  def min
    minimum = fields_products.minimum(:price)
    return 0 unless minimum
    minimum.to_i
  end

  def max
    maximum = fields_products.maximum(:price)
    return 100 unless maximum
    maximum.ceil
  end

  def products
    @filter.products.where 'price >= ? AND price <= ?', min_b, max_b
  end

  def set_filters
    @filter.filters <<
      { type: 'PriceFilter',
        name: 'price',
        params: [min_b, max_b],
        min:  min,
        max:  max,
        minB: min_b,
        maxB: max_b,
        minFormB: min_b,
        maxFormB: max_b }
  end
end
